src/renderer/components/library/SceneGenerator.tsx

```diff
@@ -0,0 +1,338 @@
+    // If we cannot parse this, default to 300
+    if (!max && max != 0) {
+      max = 300;
+    }
+
+    return (
+      <div className="SceneGenerator">
+        <div className="u-button-row">
+          <div className="u-abs-center">
+            {this.state.isEditingName && (
+              <form className="SceneNameForm" onSubmit={this.endEditingName.bind(this)}>
+                <input
+                  autoFocus
+                  type="text"
+                  ref={this.nameInputRef}
+                  value={this.props.scene.name}
+                  onBlur={this.endEditingName.bind(this)}
+                  onChange={this.updateSceneName.bind(this)}/>
+              </form>
+            )}
+            {!this.state.isEditingName && (
+              <h2
+                className="SceneGenerator__SceneGeneratorHeader u-clickable"
+                onClick={this.beginEditingName.bind(this)}>{this.props.scene.name}</h2>
+            )}
+          </div>
+
+          <div className="u-button-row-right">
+            <SimpleTextInput
+              label="Max"
+              value={this.state.max}
+              isEnabled={true}
+              onChange={this.onUpdateMax.bind(this)}/>
+            <div
+              className={`SceneGenerator__Generate u-button ${this.props.scene.sources.length > 0 ? 'u-clickable' : 'u-disabled'}`}
+              onClick={this.props.scene.sources.length > 0 ? this.previousScene.bind(this) : this.nop}>
+              Previous Scene
+            </div>
+            <div
+              className={`SceneGenerator__Generate u-button ${(sum > 0 || hasAll) && max > 0 ? 'u-clickable' : 'u-disabled'}`}
+              onClick={(sum > 0 || hasAll) && max > 0 ? this.generateScene.bind(this, tagWeights) : this.nop}>
+              Generate Scene
+            </div>
+          </div>
+          <div className="BackButton u-button u-clickable" onClick={this.props.goBack}>Back</div>
+          <div
+            className="DeleteButton u-destructive u-button u-clickable"
+            onClick={this.props.onDelete.bind(this, this.props.scene)}>
+            Delete
+          </div>
+        </div>
+
+        <div className="SceneGenerator__Content ControlGroupGroup">
+          {this.props.tags.map((tag) =>
+            <ControlGroup key={tag.id} title={tag.name} isNarrow={true}>
+              <span>{"Weight " + (Array.from(tagWeights.keys()).filter((t) => t.id == tag.id).length > 0 ?
+                tagWeights.get(Array.from(tagWeights.keys()).filter((t) => t.id == tag.id)[0]).value : 0).toString()}</span>
+              <span>{"Percentage: " + (sum > 0 ? (Array.from(tagWeights.keys()).filter((t) => t.id == tag.id).length > 0 ?
+                (tagWeights.get(Array.from(tagWeights.keys()).filter((t) => t.id == tag.id)[0]).value > 0 ?
+                  Math.round((tagWeights.get(Array.from(tagWeights.keys()).filter((t) => t.id == tag.id)[0]).value / sum) * 100) + "%" : "--") : "--") : "--")}</span>
```

Comment:
That's quite the conditional you have there, chief! :-)

Reply:
Made that a little more easily parseable